packages/flutter_tools/lib/src/commands/daemon.dart

```diff
@@ -600,51 +601,102 @@ class AppDomain extends Domain {
     }
 
-    if (_inProgressHotReload != null) {
-      throw 'hot restart already in progress';
-    }
-
-    _inProgressHotReload = app._runInZone<OperationResult>(this, () {
-      return app.restart(fullRestart: fullRestart, pause: pauseAfterRestart, reason: restartReason);
-    });
-    return _inProgressHotReload.whenComplete(() {
-      _inProgressHotReload = null;
-    });
+    return _queueAndDebounceReloadAction(
+      app,
+      fullRestart ? 'restart' : 'reload',
+      debounce,
+      () {
+        return app.restart(
+            fullRestart: fullRestart,
+            pause: pauseAfterRestart,
+            reason: restartReason);
+      },
+    );
   }
 
   Future<OperationResult> reloadMethod(Map<String, dynamic> args) async {
     final String appId = _getStringArg(args, 'appId', required: true);
     final String classId = _getStringArg(args, 'class', required: true);
-    final String libraryId =  _getStringArg(args, 'library', required: true);
+    final String libraryId = _getStringArg(args, 'library', required: true);
+    final bool debounce = _getBoolArg(args, 'debounce') ?? false;
 
     final AppInstance app = _getApp(appId);
     if (app == null) {
       throw "app '$appId' not found";
     }
 
-    if (_inProgressHotReload != null) {
-      throw 'hot restart already in progress';
+    return _queueAndDebounceReloadAction(
+      app,
+      'reloadMethod',
+      debounce,
+      () {
+        return app.reloadMethod(classId: classId, libraryId: libraryId);
+      },
+    );
+  }
+
+  /// Debounce and queue reload actions.
+  ///
+  /// Only one reload action will run at a time. Actions requested in quick
+  /// succession (within [_hotReloadDebounceDuration]) will be merged together
+  /// and all return the same result. If an action is requested after an identical
+  /// action has already started, it will be queued and run again once the first
+  /// action completes.
+  Future<OperationResult> _queueAndDebounceReloadAction(
+    AppInstance app,
+    String operationType,
+    bool debounce,
+    Future<OperationResult> Function() action,
+  ) {
+    // If there is already an operation of this type waiting to run, reset its
+    // debounce timer and return its future.
+    if (_reloadOperationQueue[operationType] != null) {
+      _reloadDebounceTimers[operationType]?.reset();
+      return _reloadOperationQueue[operationType];
     }
 
-    _inProgressHotReload = app._runInZone<OperationResult>(this, () {
-      return app.reloadMethod(classId: classId, libraryId: libraryId);
-    });
-    return _inProgressHotReload.whenComplete(() {
-      _inProgressHotReload = null;
-    });
+    // Otherwise, put one in the queue with a timer.
+    final Completer<OperationResult> completer = Completer<OperationResult>();
+    _reloadOperationQueue[operationType] = completer.future;
+    _reloadDebounceTimers[operationType] = RestartableTimer(
+      debounce ? _hotReloadDebounceDuration : Duration.zero,
```

Comment:
When would thus be null?

Reply:
`debounce`? It won't be null, but could be false. Right now the client has to opt-in to this debounce behaviour, since we don't want to result in two lots of delays for existing clients that don't know about this. If they don't send `debounce=true` we'll execute the restart/reload immediately.